org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -86,33 +88,28 @@ public void setAutoClose(boolean on) {
         autoClose = on;
     }
 
-    /**
-     * Internal helper to bound the size of a blocking read to the maximum
-     * data available.
-     */
-    private int boundRead(int suggested) throws IOException {
-        // When there's consumed data left to read, ensure we bound by the
-        // amount available there before continuing.
-        if (consumed != null && consumed.available() > 0) {
-            return Math.min(consumed.available(), suggested);
-        }
-
-        // By setting consumed = null when consumed no longer has bytes
-        // available, we provide an easy check for which channel to read
-        // from.
-        consumed = null;
-        consumedChannel = null;
-
-        // If its a non-blocking underlying socket, then return suggested;
-        // it'll read as much as currently available.
-        if (!isBlocking()) {
-            return suggested;
+    private int remoteRead() throws IOException {
+        if (consumed != null) {
+            int n = consumedChannel.read(readBuffer);
+            if (n < 0) {
```

Comment:
According to https://docs.oracle.com/javase/8/docs/api/java/nio/channels/ReadableByteChannel.html, 0 is a valid return result and we should probably set `consumed` and `consumedChannel` to `null` in both cases.

Reply:
0 is indeed a valid return value.  It means that zero bytes were read.  It doesn't mean end of stream was reached, so I think it's wrong to discard the `consumedChannel` in this case.

Based on `consumedChannel = Channels.newChannel(InputStream consumed)`, return value of 0 seems to be a "can't happen" and therefore your proposal is (vacuously) safe.  But I'd rather code to the interface than rely on these implementation details.

If you still feel strongly we sohuld change it, I will change it and add some commentary.